### PR TITLE
Enable complex autograd for `index`, add `index` and `index_put` OpInfos

### LINKED ISSUE #53605: Add complex autograd support for torch.Tensor.index
```
>>> x = torch.ones(size=(10, 10), dtype=torch.cdouble, requires_grad=True)
>>> y=x.clone()
>>> mask = torch.zeros(size=(10, ), dtype=torch.bool)
>>> y.index_put_((mask, ), y[mask], accumulate=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: index does not support automatic differentiation for outputs with complex dtype.
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @aocsa

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #53605

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -3417,6 +3463,21 @@ def reference_lgamma(x):
...
+           op=torch.Tensor.__getitem__,
+           sample_inputs_func=sample_inputs_getitem,
+           skips=(SkipInfo('TestCommon', 'test_variant_consistency_jit'),)),
+    OpInfo('index_put',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           supports_out=False,
+           supports_inplace_autograd=True,
+           sample_inputs_func=sample_inputs_index_put,
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_jit'),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3479]: why is this test being skipped? if it's intentionally skipped due to a test failure, could you share the error message or the reason why it's failing?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -3417,6 +3463,21 @@ def reference_lgamma(x):
            dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
            supports_out=False,
            sample_inputs_func=sample_inputs_index_add),
+    OpInfo('__getitem__',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           supports_out=False,
+           supports_inplace_autograd=False,
+           op=torch.Tensor.__getitem__,
+           sample_inputs_func=sample_inputs_getitem,
+           skips=(SkipInfo('TestCommon', 'test_variant_consistency_jit'),)),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3472]: same comment as below

**Review by anjali411 — APPROVED** [review]
thanks @kurtamohler ! This PR lgtm overall. I just have one question regarding the skip, but besides that this PR is ready to be merged.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_methods_invocations.py:3479]: Yeah there's a failure that I don't understand yet:

<details>
<summary>click to expand</summary>

```
======================================================================
ERROR: test_variant_consistency_jit_index_put_cpu_float32 (__main__.TestCommonCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 297, in instantiated_test
    raise rte
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/test/test_ops.py", line 294, in test_variant_consistency_jit
    check_against_reference(self,
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_jit.py", line 63, in check_against_reference
    outputs_test = self.runAndSaveRNG(func, nograd_inputs, kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_jit.py", line 130, in runAndSaveRNG
    results = func(*inputs, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/jit_metaprogramming_utils.py", line 301, in script_fn
    fn, tensors = gen_script_fn_and_args(method_name, func_type, *args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/jit_metaprogramming_utils.py", line 292, in gen_script_fn_and_args
    CU = torch.jit.CompilationUnit(script)
RuntimeError: 
undefined value tensor:
  File "<string>", line 3

def the_method(i0, i1):
    return torch.index_put(i0, (tensor([4, 4]),), i1, accumulate=False)
```

</details>

Seems like it must be an existing issue that this PR did not introduce, but I could look into it anyway

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_methods_invocations.py:3472]: This is the failure for this one:

<details>
<summary>click to expand</summary>

```
======================================================================
ERROR: test_variant_consistency_jit_index_put_cuda_float32 (__main__.TestCommonCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_utils.py", line 955, in wrapper
    method(*args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_utils.py", line 955, in wrapper
    method(*args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 297, in instantiated_test
    raise rte
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/test/test_ops.py", line 294, in test_variant_consistency_jit
    check_against_reference(self,
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_jit.py", line 63, in check_against_reference
    outputs_test = self.runAndSaveRNG(func, nograd_inputs, kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/common_jit.py", line 130, in runAndSaveRNG
    results = func(*inputs, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/jit_metaprogramming_utils.py", line 301, in script_fn
    fn, tensors = gen_script_fn_and_args(method_name, func_type, *args, **kwargs)
  File "/work2/kurtamohler/development/pytorch-index_put-complex-autograd/torch/testing/_internal/jit_metaprogramming_utils.py", line 292, in gen_script_fn_and_args
    CU = torch.jit.CompilationUnit(script)
RuntimeError: 
undefined value tensor:
  File "<string>", line 3

def the_method(i0, i1):
    return torch.index_put(i0, (tensor([0, 1], device='cuda:0'),), i1, accumulate=False)
                                ~~~~~~ <--- HERE

```

</details>

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:3472]: no idea how to fix this off the top of my head. let's follow up on this in a separate issue/subsequent PR?

**Comment by anjali411** [comment]
@kurtamohler could you please rebase?

**Comment by anjali411** [comment]
@kurtamohler looks like another PR went in before this one. needs another rebase!

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:3514]: did you get a chance to look into this?

**Comment by kurtamohler** [comment]
@anjali411, rebase done

**Pull request closed without merging**